### PR TITLE
[EPLB] Add offline expert mapping support (aka "offline EPLB")

### PR DESCRIPTION
**Pull request by arpera**
## Purpose

Adds offline EPLB support: collect expert-load stats during a representative run, save them to a file, then reload them at startup of a different server to apply a static physical-to-logical mapping before warmup.

### Performance

Hardware: 8xB200, model `nvidia/Qwen3.5-397B-A17B-NVFP4` (DP=8 EP).

For each dataset the corpus is split 50/50 into two parts. Schedule is recorded on `part1` and perf is measured on `part2` — independent data.

| Dataset (size after 50/50 split) | Baseline | Online EPLB | Offline EPLB | Δ Total token throughput (tok/s) |
| --- | --- | --- | --- | --- |
| [AI4Chem/ChemData700K](https://huggingface.co/datasets/AI4Chem/ChemData700K) (~324k/half) | 6745.20 | 6873.39 | 6981.33 | +3.5% |
| [nvidia/OpenMathInstruct-2](https://huggingface.co/datasets/nvidia/OpenMathInstruct-2) (~303k/half) | 6473.46 | — | 6717.14 | +3.8% |
| [ise-uiuc/Magicoder-Evol-Instruct-110K](https://huggingface.co/datasets/ise-uiuc/Magicoder-Evol-Instruct-110K) (~55k/half) | 6381.60 | — | 6598.05 | +3.4% |

<details>
<summary>Bench commands</summary>

Shared server flags: `-tp 1 -pp 1 -dp 8 --enable-expert-parallel --language-model-only --reasoning-parser qwen3 --stream-interval 100 --enable-eplb`. Only the `--eplb-config` and the per-dataset bench arguments differ.

```
# Step 1: collect stats on part1 (also doubles as the baseline reference
# for math / code, since `save_path` records stats but does not apply any
# rearrangement)
vllm serve nvidia/Qwen3.5-397B-A17B-NVFP4 --port 8000 [shared flags] \
    --eplb-config '{"save_path":"./cumulative.safetensors","step_interval":256,"communicator":"torch_gloo"}'

# Step 1 (chemistry only): separate no-schedule baseline server on part2
vllm serve nvidia/Qwen3.5-397B-A17B-NVFP4 --port 8000 [shared flags] \
    --eplb-config '{"step_interval":100000,"communicator":"torch_gloo"}'

# Step 2: offline EPLB server (schedule loaded, online rearrange pinned off)
vllm serve nvidia/Qwen3.5-397B-A17B-NVFP4 --port 8000 [shared flags] \
    --eplb-config '{"load_path":"./cumulative.safetensors","step_interval":100000,"communicator":"torch_gloo"}'

# Client (3 runs each phase, paired --seed across baseline/offline)
vllm bench serve --backend vllm --model nvidia/Qwen3.5-397B-A17B-NVFP4 \
    --port 8000 --endpoint /v1/completions --temperature 0.0 \
    --dataset-name custom --dataset-path ./<dataset>/part<N>.jsonl \
    --custom-output-len 2048 --skip-chat-template \
    --num-prompt <NUM_PROMPT> --max-concurrency 128 \
    --seed <random per iter, same value in baseline and offline>

# Per-dataset bench params:
#   chemistry: num-prompt 1024
#   math:      num-prompt 1024
#   code:      num-prompt 512  (sized down so a single run finishes in ~2 min)
```

</details>

<details>

```bash
vllm serve nvidia/Qwen3.5-397B-A17B-NVFP4 --port 8000 -tp 1 -pp 1 -dp 8 --enable-expert-parallel --language-model-only --reasoning-parser qwen3 --stream-interval 100 --enable-eplb --eplb-config '{"step_interval":1000, "communicator":"nixl"}'

vllm bench serve --backend vllm --model nvidia/Qwen3.5-397B-A17B-NVFP4 --port 8000 --endpoint /v1/completions --temperature 0.0 --dataset-name custom --dataset-path chemdata/part2.jsonl --custom-output-len 2048 --skip-chat-template --num-prompt 1024 --max-concurrency 128
```

</details>

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/distributed/eplb/eplb_state.py [hunk 1]
```diff
@@ -339,6 +350,146 @@ def validate_ep_configuration(self, new_model: MixtureOfExperts):
...
+        self,
+        mapping_path: str,
+        num_moe_layers: int,
+        num_physical_experts: int,
+        num_logical_experts: int,
+        num_redundant_experts: int,
+        max_slots_per_logical: int,
+    ) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """Load per-layer physical-to-logical mapping from a JSONL file."""
+        cfg = self._load_initial_mapping_record(mapping_path)
```
Comment (ilmarkov) [inline vllm/distributed/eplb/eplb_state.py]: We should validate that the mapping is the same on all ranks.

File: vllm/distributed/eplb/eplb_state.py [hunk 2]
```diff
@@ -369,48 +510,85 @@ def add_model(
...
-                -1,
+            # CORRECTNESS: the model has just been loaded with weights in
+            # identity order (phys slot i holds logical expert i, plus
+            # redundant copies of the first few experts). The JSONL mapping above
+            # swapped the maps but did NOT move the physical weights, so
+            # the router would end up dispatching tokens for logical X to
+            # the slot of some other logical Y. Defer the physical
+            # rearrangement to the first profile step, where NCCL is
+            # ready and the kernel already exists.
+            self._pending_initial_mapping_rearrange = True
```
Comment (ilmarkov) [inline vllm/distributed/eplb/eplb_state.py]: It should be per-model. As it guards per-model data.

Comment (ilmarkov) [reply vllm/distributed/eplb/eplb_state.py]: Save it in `EplbModelState`

File: vllm/distributed/eplb/eplb_state.py [hunk 3]
```diff
@@ -470,6 +648,87 @@ def add_model(
...
+        """Append one JSONL record with expert-load statistics."""
+        if num_tokens_per_rank.sum().item() == 0:
+            return
+
+        if self._expert_load_stats_file is None:
+            stats_path = self.parallel_config.eplb_config.expert_load_stats_path
+            assert stats_path is not None
+            path = Path(stats_path)
+            path.parent.mkdir(parents=True, exist_ok=True)
+            self._expert_load_stats_file = open(path, "a")  # noqa: SIM115
```
Comment (ilmarkov) [inline vllm/distributed/eplb/eplb_state.py]: Needs to be closed somewhere. Potential leak at elastic changes.

File: vllm/distributed/eplb/eplb_state.py [hunk 4]
```diff
@@ -554,6 +820,29 @@ def step(
                         - self.expert_rearrangement_step,
                     )
 
+        should_write_load_stats = (
+            self.parallel_config.eplb_config.expert_load_stats_path is not None
+        )
+        if should_write_load_stats:
```
Comment (ilmarkov) [inline vllm/distributed/eplb/eplb_state.py]: It seems to conflict with #39178

Comment (ilmarkov) [reply vllm/distributed/eplb/eplb_state.py]: It shouldn't do sync and write to a file at each step. We might use `expert_load_stats_path` at the actual runtime not only for profiling.

File: tools/eplb/generate_static_mapping.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    parser.add_argument("--num-redundant-experts", type=int, required=True)
+    parser.add_argument("--start-step", type=int, default=None)
+    parser.add_argument("--end-step", type=int, default=None)
+    parser.add_argument(
+        "--num-ranks",
+        type=int,
+        default=None,
+        help="Number of EP ranks. Defaults to the num_ranks field in the stats.",
+    )
+    parser.add_argument("--num-groups", type=int, default=1)
```
Comment (ilmarkov) [inline tools/eplb/generate_static_mapping.py]: Probably, it should be saved in the jsonl. `num_groups` is critical for DeepSeek models.

Comment (ilmarkov) [reply tools/eplb/generate_static_mapping.py]: Or taken from model config.

File: vllm/distributed/eplb/eplb_state.py [hunk 6]
```diff
@@ -470,6 +648,87 @@ def add_model(
         self.model_states[model_config.compute_hash()] = model_state
         self.num_valid_physical_experts = model.num_physical_experts
 
+    def _apply_initial_mapping_rearrange(self, ep_group: ProcessGroup) -> None:
```
Comment (ilmarkov) [inline vllm/distributed/eplb/eplb_state.py]: Why not to re-use the existing primitives that do the rearrangment? `rearrange_expert_weights_inplace` should work

File: vllm/distributed/eplb/eplb_state.py [hunk 7]
```diff
@@ -498,6 +757,13 @@ def step(
         ep_group = get_ep_group().device_group
         if is_profile:
             self.rearrange(is_profile=True)
+            # Static initial mapping: physically move expert weights to
+            # match the JSONL mapping that was loaded in add_model. Only
+            # runs once (guarded by the flag), right after the dummy
+            # rearrange set up the NCCL buffers.
+            if self._pending_initial_mapping_rearrange:
+                self._apply_initial_mapping_rearrange(ep_group)
```
Comment (ilmarkov) [inline vllm/distributed/eplb/eplb_state.py]: I don't think we want to do heavy memory movements (especially if doesn't make sense with `disable_online_rebalancing=True`) at profile. It will reduce KV cache budget

Comment (ilmarkov) [reply vllm/distributed/eplb/eplb_state.py]: I think start-up only operation shouldn't be in profile_run.

**Review by ilmarkov — CHANGES_REQUESTED** [review]
Thank you for the PR!

Added the comments. Major issues:
1. We don't want to sync and dump steps at every step when we dump is enabled.
2. The moment when we do initial memory movements has to be out of profile.
3. Also, it needs to be verified how it behaves with elastic ep.

Could you add experiments with non-random data?

File: vllm/distributed/eplb/eplb_state.py (hunk 1 above)
Comment (arpera) [reply vllm/distributed/eplb/eplb_state.py]: Done

File: vllm/distributed/eplb/eplb_state.py (hunk 2 above)
Comment (arpera) [reply vllm/distributed/eplb/eplb_state.py]: Done

File: vllm/distributed/eplb/eplb_state.py (hunk 7 above)
Comment (arpera) [reply vllm/distributed/eplb/eplb_state.py]: Done

File: vllm/distributed/eplb/eplb_state.py (hunk 4 above)
Comment (arpera) [reply vllm/distributed/eplb/eplb_state.py]: Fixed

File: vllm/distributed/eplb/eplb_state.py (hunk 3 above)
Comment (arpera) [reply vllm/distributed/eplb/eplb_state.py]: Fixed

File: tools/eplb/generate_static_mapping.py (hunk 5 above)
Comment (arpera) [reply tools/eplb/generate_static_mapping.py]: Done

File: vllm/distributed/eplb/eplb_state.py (hunk 6 above)
Comment (arpera) [reply vllm/distributed/eplb/eplb_state.py]: Done

**Comment by vadiklyutiy** [comment]
@arpera and I had a conversation. Seems it would be better to save expert stat only and during model loading make a decision about expert placing. It allows to use one stat file for model and any TP* will use the file.

**Review by vadiklyutiy — COMMENTED** [review]
@arpera Could you pls add some short doc how to use it

File: vllm/v1/worker/gpu_model_runner.py [hunk 8]
```diff
@@ -3203,9 +3096,12 @@ def eplb_step(self, is_dummy: bool = False, is_profile: bool = False) -> None:
         """
         if not self.parallel_config.enable_eplb or self.eep_eplb_suppressed:
             return
+        if self.eplb_state is None:
```
Comment (SageMoore) [inline vllm/v1/worker/gpu_model_runner.py]: Did you find a case where we pass the `if not self.parallel_config.enable_eplb` check but don't have an eplb_state?

File: vllm/v1/worker/gpu_model_runner.py [hunk 9]
```diff
@@ -5104,7 +4892,17 @@ def load_model(self, load_dummy_weights: bool = False) -> None:
...
         if (
-            self._moe_model is not None
+            is_mixture_of_experts(self.model)
+            and self.parallel_config.enable_eplb
+            and not load_dummy_weights
+            and self.eplb_state is not None
+            and self.parallel_config.eplb_config.load_path is not None
+        ):
+            self.eplb_state.rearrange(load_initial=True)
+            self.eplb_state = None
```
Comment (SageMoore) [inline vllm/v1/worker/gpu_model_runner.py]: Why do you set this to None here?

File: vllm/distributed/eplb/eplb_state.py [hunk 10]
```diff
@@ -687,35 +696,67 @@ def rearrange(
...
-                eplb_model_state.model.num_moe_layers,
-                eplb_model_state.model.num_logical_experts,
-                dtype=eplb_model_state.expert_load_window.dtype,
-                device=eplb_model_state.expert_load_window.device,
-            )
-            logical_expert_load_window.scatter_add_(
-                dim=-1,
-                index=eplb_model_state.physical_to_logical_map[
-                    :, : self.num_valid_physical_experts
+        if load_initial:
```
Comment (SageMoore) [inline vllm/distributed/eplb/eplb_state.py:757]: Instead of doing this all here let's add a new function: `global_expert_load_windows.append( compute_global_expert_load_window(eplb_model_state, load_initial, ...))`


Let's also add the following comment
```
# Uncommon Case: If the user has provided an expert load window for initial placement, 
# load it instead of computing it.
if load_initial:

File: vllm/distributed/eplb/eplb_state.py [hunk 11]
```diff
@@ -687,35 +696,67 @@ def rearrange(
...
-            global_expert_load_window = logical_expert_load_window.sum(dim=0)
-            global_expert_load_windows.append(global_expert_load_window)
+                global_expert_load_window = logical_expert_load_window.sum(dim=0)
+                global_expert_load_windows.append(global_expert_load_window)
         # Perform all-reduce to get the expert load across all ranks for each model
         global_expert_load_windows = self._allreduce_list(global_expert_load_windows)
 
+        save_path = self.parallel_config.eplb_config.save_path
+        if save_path is not None and not is_profile:
+            for (name, _), global_load in zip(
```
Comment (SageMoore) [inline vllm/distributed/eplb/eplb_state.py]: Same idea here
```
if save_path is not None ...:
    write_load_to_file(save_path)
    return None
```

File: vllm/distributed/eplb/eplb_state.py [hunk 12]
```diff
@@ -687,35 +696,67 @@ def rearrange(
...
+                load_cpu = global_load.detach().to(dtype=torch.float32, device="cpu")
+                if name in self.cumulative_logical_load:
+                    self.cumulative_logical_load[name].add_(load_cpu)
+                else:
+                    self.cumulative_logical_load[name] = load_cpu.clone()
+            if get_ep_group().device_group.rank() == 0:
+                self._save_logical_load(self.cumulative_logical_load, Path(save_path))
+                logger.info("Saved EPLB cumulative logical load to %s.", save_path)
+            # Recording-only mode: skip physical rearrangement so we capture
+            # the baseline on the unchanged topology and avoid NCCL p2p OOM
```
Comment (SageMoore) [inline vllm/distributed/eplb/eplb_state.py]: I don't fully understand this comment? Is this just to get around https://github.com/vllm-project/vllm/issues/38986? IMO the solution there is to use NIXL by default when it's installed and fallback to GLOO otherwise.

**Review by SageMoore — COMMENTED** [review]
Hi, @arpera. Thanks for the contribution. I have two high level pieces of feedback for this PR.

First, just to give some more context, we've been working on making Async EPLP the most performant and low-overhead way to run EPLB for a while now. With the NIXL backend, we've been able to get rid of the numerous hangs and overlapping the expert transfers with the forward pass has gone a long way towards making EPLB cheaper to run. Frankly there's a bit more work to do here, but the main benefit of a more performant EPLB is that you can run it more frequently increasing the balancedness, which increases the overall throughput.

It's not clear to me that the approach here is the best way forward. The idea of an informed initial placement makes sense, and the benchmarks you posted do show a decent performance improvement. However, if we ran EPLB after some small number of initial steps I suspect that we could probably get similar results with lower user friction. EPLB is primarily a Wide EP feature, meaning pushing/pulling files is a bit more cumbersome. Not impossible obviously, and definitely automatable, but still there's friction. I'm interested to hear your thoughts on this.

Second, given that this is a somewhat niche feature, let's try to sequester the code for it as much as possible. I've left some comments on how I think that could be done.

File: vllm/distributed/eplb/eplb_state.py (hunk 12 above)
Comment (arpera) [reply vllm/distributed/eplb/eplb_state.py]: Yes, this comment was added during debugging process on the same issue that I reported in #38986, I will remove this comment it makes no sense here now. Thanks for highlighting this.

Comment (SageMoore) [reply vllm/distributed/eplb/eplb_state.py]: Just a heads up @ilmarkov sent out a PR that at least prevents the default case from hanging with multi-weight models https://github.com/vllm-project/vllm/pull/43110. IMO the two NCCL backends will become harder to justify once https://github.com/vllm-project/vllm/pull/41633 merges. If we can close the performance gap with NIXL, then we will likely deprecate them.

**Comment by arpera** [comment]
Thank you @SageMoore for a detailed review!

I agree that on practice the difference between offline and online EPLB is likely to be small, and that online EPLB requires fewer intermediate steps. My motivation for offline EPLB is based on the following.

First, offline and online EPLB modes are orthogonal to each other, which means they can be used either independently or together. In other words, offline EPLB is not a replacement for online EPLB. Offline EPLB complements online EPLB and can be used by users who understand their inference workload pattern well enough.

Second, offline EPLB has one important advantage over online EPLB: it does not require additional communications between ranks during inference. This is beneficial because it improves the overall robustness of the system, since hard bugs to detect and debug usually happen during all2all communications. In case of distributed serving, robustness becomes even more important, because in such systems anything can go wrong, and that is normal. In this case reducing amount of communications in inference makes sense for me.

File: vllm/distributed/eplb/eplb_state.py (hunk 10 above)
Comment (arpera) [reply vllm/distributed/eplb/eplb_state.py:757]: Good suggestion. Done.

File: vllm/v1/worker/gpu_model_runner.py (hunk 8 above)
Comment (arpera) [reply vllm/v1/worker/gpu_model_runner.py]: This I did intentionally to save memory in case we do not need to collect statistics, so if we only use offline EPLB mode. I added a new EPLB config parameter `disable_offline` that is False by default. Probably now it should be more clear. Have a look into last commit.

File: vllm/v1/worker/gpu_model_runner.py (hunk 9 above)
Comment (arpera) [reply vllm/v1/worker/gpu_model_runner.py]: I set it there to None to disable further rebalancing when I use offline EPLB. I updated the patch and introduced a new EPLB config parameter `disable_offline` that should help there.

File: vllm/distributed/eplb/eplb_state.py (hunk 11 above)
Comment (arpera) [reply vllm/distributed/eplb/eplb_state.py]: Done

File: vllm/distributed/eplb/eplb_state.py (hunk 12 above)
Comment (arpera) [reply vllm/distributed/eplb/eplb_state.py]: Done
